File: langs/python/src/vmx/capabilities/dialog.py
"""Dialog / form capability contracts. See spec/14-capabilities.md §Dialog / form."""

from __future__ import annotations

from abc import ABC, abstractmethod


class IClosable(ABC):
    @abstractmethod
    def can_close(self) -> bool: ...

    @abstractmethod
    def close(self) -> None: ...


class IApprovable(ABC):
    @abstractmethod
    def can_approve(self) -> bool: ...

    @abstractmethod
    def approve(self) -> None: ...


class ICancelable(ABC):
    @abstractmethod
    def can_cancel(self) -> bool: ...

    @abstractmethod
    def cancel(self) -> None: ...

File: langs/csharp/src/VMx/Capabilities/Dialog.cs
namespace VMx.Capabilities;

// Capability interfaces from spec/14-capabilities.md §Dialog / form.

/// <summary>Capability: the implementer can be closed.</summary>
public interface IClosable
{
    /// <summary>Returns true when <see cref="Close"/> is valid to call.</summary>
    bool CanClose();

    /// <summary>Performs the close action.</summary>
    void Close();
}

/// <summary>Capability: the implementer can be approved.</summary>
public interface IApprovable
{
    /// <summary>Returns true when <see cref="Approve"/> is valid to call.</summary>
    bool CanApprove();

    /// <summary>Performs the approve action.</summary>
    void Approve();
}

/// <summary>Capability: the implementer can be canceled.</summary>
public interface ICancelable
{
    /// <summary>Returns true when <see cref="Cancel"/> is valid to call.</summary>
    bool CanCancel();

    /// <summary>Performs the cancel action.</summary>
    void Cancel();
}
